File: src/main/java/eu/happybe/openapi/servers/ServerGroup.java
package eu.happybe.openapi.servers;

import cn.nukkit.Player;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ServerGroup {

    @Getter
    private final String groupName;
    @Getter
    private final List<Server> servers = new ArrayList<>();

    public ServerGroup(String groupName) {
        this.groupName = groupName;
    }

    public boolean canAddServer(Server server) {
        return server.getServerName().contains(this.getGroupName());
    }

    public void addServer(Server server) {
        this.getServers().add(server);
    }

    public int getOnlinePlayers() {
        int online = 0;
        for(Server server : this.getServers()) {
            online += server.getOnlinePlayers();
        }

        return online;
    }

    public Server getFitServer(Player player) {
        List<Server> servers = this.getServers().stream()
                .filter(server -> !server.getServerName().equals(ServerManager.getCurrentServer().getServerName()))
                .filter(Server::isOnline)
                .filter(server -> (!server.isWhitelisted()) || player.hasPermission("happybe.operator"))
                .sorted((firstServer, secondServer) -> Integer.compare(secondServer.getOnlinePlayers(), firstServer.getOnlinePlayers()))
                .collect(Collectors.toList());

        if(servers.size() > 0) {
            return servers.get(0);
        }

        return null;
    }
}

File: src/main/java/eu/happybe/openapi/servers/ServerGroup.kt
package eu.happybe.openapi.servers

import cn.nukkit.Player
import lombok.Getter
import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors

class ServerGroup(@field:Getter private val groupName: String) {
    @Getter
    private val servers: List<Server> = ArrayList()
    fun canAddServer(server: Server): Boolean {
        return server.getServerName().contains(this.getGroupName())
    }

    fun addServer(server: Server?) {
        this.getServers().add(server)
    }

    val onlinePlayers: Int
        get() {
            var online = 0
            for (server in this.getServers()) {
                online += server.getOnlinePlayers()
            }
            return online
        }

    fun getFitServer(player: Player): Server? {
        val servers: List<Server> = this.getServers().stream()
                .filter(Predicate { server: Server -> server.getServerName() != ServerManager.getCurrentServer().getServerName() })
                .filter(Predicate { obj: Server -> obj.isOnline() })
                .filter(Predicate { server: Server -> !server.isWhitelisted() || player.hasPermission("happybe.operator") })
                .sorted(java.util.Comparator { firstServer: Server, secondServer: Server -> Integer.compare(secondServer.getOnlinePlayers(), firstServer.getOnlinePlayers()) })
                .collect(Collectors.toList())
        return if (servers.size > 0) {
            servers[0]
        } else null
    }
}